### ISSUE TITLE
[BUG] L-BFGS IK uses a static initial inverse-Hessian scale, so floating-base convergence degrades with distance

### ISSUE THREAD

**Issue by eric-heiden** [newton-physics/newton#4090]
## Bug Description

`IKOptimizerLBFGS` uses a **fixed scalar** for the initial inverse-Hessian approximation `H0`. In the two-loop recursion (`_compute_search_direction_template` in `newton/_src/sim/ik/ik_lbfgs_optimizer.py`), the entire vector is scaled by one constant:

```python
# Apply initial Hessian approximation in-place
for j in range(DOF):
    q[j] = h0_scale * q[j]
```

`h0_scale` is a single Python float (default `1.0`) passed straight through as a kernel scalar. It is the same for every batch row and every iteration, and it is never derived from the `(s, y)` correction history.

Standard L-BFGS instead uses a dynamic scale,

```
H0_k = gamma_k * I,   gamma_k = (s_{k-1} . y_{k-1}) / (y_{k-1} . y_{k-1})
```

(Nocedal and Wright, *Numerical Optimization*, eq. 7.20), which tracks the local curvature scale automatically, per problem and per iteration.

The LM optimizer already has an equivalent mechanism: `IKOptimizerLM` stores `lambda_values` as a **per-row array** and adapts it every iteration within `[lambda_min, lambda_max]`. L-BFGS has no comparable per-row, per-iteration knob.

### Why this shows up on floating bases

For a `FREE` or `DISTANCE` root joint, the angular columns of the objective Jacobian scale with the distance from the joint's parent anchor to the controlled point. Translate a floating-base robot 100 m from its anchor and those angular columns become roughly 100x larger than the linear ones, while `H0 = 1.0 * I` still weights every DOF identically. The resulting search direction is badly scaled, and the line search cannot repair it: it picks from a fixed candidate list (`[0.1, 0.2, 0.5, 0.8, 1.0, 1.5, 2.0, 3.0]`), so it can only shrink the step, never rebalance the direction between angular and linear DOFs.

Net effect: L-BFGS accuracy degrades with base distance while LM stays flat.

### Measurements

Floating base (`FREE` root) plus a 2-link arm, one position objective, 32 iterations, `ANALYTIC` Jacobian. Final tip position error in meters, measured on top of the corrected FREE/DISTANCE Jacobians from #4088:

| optimizer | 0 m | 1 m | 5 m | 20 m | 50 m | 100 m |
|---|---|---|---|---|---|---|
| LM (adaptive damping) | 9.9e-07 | 3.0e-07 | 4.8e-07 | 0.0 | 3.8e-06 | 7.6e-06 |
| **L-BFGS `h0_scale=1.0`** (default) | 7.3e-07 | 5.3e-07 | 4.8e-07 | **9.5e-04** | **8.2e-02** | **3.1e+00** |
| L-BFGS `h0_scale=0.1` | 3.5e-07 | 3.7e-05 | 1.2e-07 | 1.9e-06 | 4.9e-06 | 1.3e-04 |
| L-BFGS `h0_scale=0.01` | 2.5e-07 | 1.2e-07 | 9.9e-07 | 0.0 | 3.8e-06 | 4.0e-05 |
| L-BFGS `h0_scale=0.001` | 2.2e-05 | 3.9e-04 | 1.8e-07 | 7.2e-07 | 2.0e-05 | 1.0e-05 |
| L-BFGS `h0_scale=0.0001` | 6.5e-06 | 6.4e-06 | 3.1e-04 | 1.2e-04 | 4.6e-05 | 6.1e-05 |

Two things stand out:

1. At the default `h0_scale=1.0` the solve fails outright at long range (3.1 m of residual error at 100 m).
2. **No single static value is good everywhere.** `1.0` is best near the origin, `0.01` is far better at distance, and `0.0001` is mediocre everywhere. The right value is a function of problem scale, which is exactly what a static parameter cannot express.

### A dynamic scale fixes it

I prototyped `gamma_k` in the two-loop recursion, computed per row from the most recent accepted `(s, y)` pair, with `h0_scale` kept as a relative multiplier:

```python
gamma = h0_scale
if count > 0:
    last = (start + count - 1) % M_HIST
    y_l = wp.tile_load(y_history[row, last], shape=(DOF,), storage="shared")
    s_l = wp.tile_load(s_history[row, last], shape=(DOF,), storage="shared")
    sy_t = wp.tile_sum(wp.tile_map(wp.mul, s_l, y_l))
    yy_t = wp.tile_sum(wp.tile_map(wp.mul, y_l, y_l))
    if yy_t[0] > 1.0e-12:
        gamma = h0_scale * sy_t[0] / yy_t[0]

for j in range(DOF):
    q[j] = gamma * q[j]
```

Same benchmark, both runs at the default `h0_scale=1.0`:

| base offset | static `H0` (current) | dynamic `gamma_k` (prototype) |
|---|---|---|
| 0 m | 7.3e-07 | 1.9e-07 |
| 1 m | 5.3e-07 | 1.2e-06 |
| 5 m | 4.8e-07 | 0.0 |
| 20 m | 9.5e-04 | 2.7e-04 |
| 50 m | 8.2e-02 | **1.3e-05** |
| 100 m | 3.1e+00 | **1.1e-04** |

Roughly 6,000x better at 50 m and 29,000x better at 100 m, with no regression at short range. The default `h0_scale=1.0` becomes a good choice at every distance, which is the point.

### Suggested fix

1. Compute `gamma_k` per row, per iteration from the most recent accepted `(s, y)` pair and use it for `H0`, keeping `h0_scale` as a relative multiplier so the default stays `1.0` and existing tuning still means something. `_update_history_template` already computes `y . s` for the curvature check, so only `y . y` is missing.
2. If the angular vs linear anisotropy still hurts after step 1, consider a diagonal `H0` (per-DOF scaling from Jacobian column norms) rather than a single scalar. A scalar `gamma_k` corrects overall magnitude but not the relative weighting between DOFs.

Worth checking whether the fixed line-search candidate list should adapt too, though the measurements suggest `H0` is the dominant term.

### Notes

- This is a pre-existing L-BFGS limitation, not a regression from #4088. On `main` the same sweep is erratic in both directions because the analytic FREE/DISTANCE Jacobians are still wrong there; correcting them is what turns this into a clean, monotonic, diagnosable scaling problem.
- One honest caveat: correcting the pivot to the parent anchor is right, and it does legitimately widen the dynamic range of the free-joint angular columns. LM absorbs that through its adaptive damping. L-BFGS currently has nothing to absorb it with, which is the gap this issue tracks.
- The prototype above is a proof of concept validated only on this benchmark. It needs the full IK test suite and the wider IK benchmark set before it becomes a real change.

## Reproduction Script

```python
import numpy as np
import warp as wp

import newton
import newton.ik as ik
from newton._src.sim.ik.ik_common import eval_fk_batched

OFFSETS = (0.0, 1.0, 5.0, 20.0, 50.0, 100.0)
ITERS = 32


def build():
    b = newton.ModelBuilder()
    base = b.add_link(xform=wp.transform([0.0, 0.0, 0.0], wp.quat_identity()), mass=1.0)
    j0 = b.add_joint_free(parent=-1, child=base)
    l1 = b.add_link(xform=wp.transform([0.5, 0.0, 0.0], wp.quat_identity()), mass=1.0)
    j1 = b.add_joint_revolute(
        parent=base, child=l1,
        parent_xform=wp.transform([0.0, 0.0, 0.0], wp.quat_identity()),
        child_xform=wp.transform([-0.5, 0.0, 0.0], wp.quat_identity()),
        axis=[0.0, 0.0, 1.0],
    )
    l2 = b.add_link(xform=wp.transform([1.5, 0.0, 0.0], wp.quat_identity()), mass=1.0)
    j2 = b.add_joint_revolute(
        parent=l1, child=l2,
        parent_xform=wp.transform([0.5, 0.0, 0.0], wp.quat_identity()),
        child_xform=wp.transform([-0.5, 0.0, 0.0], wp.quat_identity()),
        axis=[0.0, 0.0, 1.0],
    )
    b.add_articulation([j0, j1, j2])
    return b.finalize(requires_grad=True), l2


def run(model, tip, optimizer, h0_scale=1.0):
    n = len(OFFSETS)
    # Target sits 0.8 m from the translated base, so it is always exactly
    # reachable: the base itself is free.
    tgt_np = np.array([[d + 0.8, 0.5, 0.0] for d in OFFSETS], dtype=np.float32)
    targets = wp.array(tgt_np, dtype=wp.vec3)

    kw = {}
    if optimizer == ik.IKOptimizer.LBFGS:
        kw["h0_scale"] = h0_scale
    solver = ik.IKSolver(
        model, n, [ik.IKObjectivePosition(tip, wp.vec3(), targets)],
        optimizer=optimizer, jacobian_mode=ik.IKJacobianType.ANALYTIC, **kw,
    )

    q = np.zeros((n, model.joint_coord_count), dtype=np.float32)
    q[:, 0] = OFFSETS  # base x translation
    q[:, 6] = 1.0      # base quaternion w
    q_in = wp.array(q, dtype=wp.float32)
    q_out = wp.zeros_like(q_in)
    solver.step(q_in, q_out, iterations=ITERS)
    wp.synchronize()

    body_q = wp.zeros((n, model.body_count), dtype=wp.transform)
    body_qd = wp.zeros((n, model.body_count), dtype=wp.spatial_vector)
    qd = wp.zeros((n, model.joint_dof_count), dtype=wp.float32)
    eval_fk_batched(model, q_out, qd, body_q, body_qd)
    wp.synchronize()
    return np.linalg.norm(body_q.numpy()[:, tip, :3] - tgt_np, axis=1)


def main():
    wp.init()
    model, tip = build()
    print("final tip position error [m] vs base offset")
    print("offset:        " + "".join("%11.0f m" % d for d in OFFSETS))
    e = run(model, tip, ik.IKOptimizer.LM)
    print("%-15s" % "LM" + "".join("%13.2e" % v for v in e))
    for h0 in (1.0, 1e-1, 1e-2, 1e-3, 1e-4):
        e = run(model, tip, ik.IKOptimizer.LBFGS, h0_scale=h0)
        print("%-15s" % ("LBFGS h0=%g" % h0) + "".join("%13.2e" % v for v in e))


if __name__ == "__main__":
    main()
```

## System Information

- Newton: 1.6.0.dev0 (with #4088 applied)
- Warp: 1.17.0.dev20260807
- Python: 3.12.12
- OS: Linux 5.15.0 (glibc 2.35)
- GPU: NVIDIA L40 (sm_89), CUDA 12.9
